CONTRIBUTING.md

```diff
@@ -47,6 +47,61 @@ and additionally the following rules:
 * Use `my ($self) = @_;` for parameter parsing in methods when accessing the
   `$self` object. Do not parse any parameter if you do not need any.
 * [DRY](https://en.wikipedia.org/wiki/Don't_repeat_yourself)
+* Avoid `sleep()`: Do not use `sleep()` or simulate a sleep-like functionality.
+  This is introducing the risk of either unnecessarily wasting time or just
+  shifting race conditions and making them even harder to investigate. You *can*
+  use sleep but there are only very limited cases where is the best solution
+  (e.g. yield threads in low-level system code). Here you have the classical
+  problem in software design of
+  [Synchronization](https://en.wikipedia.org/wiki/Synchronization_(computer_science))
+  . sleep should not be used in place of proper synchronization methods.
+  https://blogs.msmvps.com/peterritchie/2007/04/26/thread-sleep-is-a-sign-of-a-poorly-designed-program/
+  is one of multiple references that explains this: "The original problem is
+  likely a timing/synchronization issue, ignoring it by hiding it with
+  Thread.Sleep is only going to delay the problem and make it occur in random,
+  hard to reproduce ways." And if the problem does not occur at least time is
+  wasted.
+  We are waiting because some condition changes some time ... keyword(s) is/are
+  *some time*. It is important to know the exact condition that the code should
+  wait for and check for that instead of delaying program execution for an
+  arbitrary amount of time. For openQA tests in general just try to do what you
+  as a human user would also do. You would not look on your watch waiting for
+  exactly 100s before you look at the screen again, right? :)
+* Avoid `is_tumbleweed()`: Our tests should always consider openSUSE Tumbleweed
+  as the default product without a version. That means any differing behaviour
+  should have an explicit exclude rule for the "older products" and potentially
+  an explanation why the behaviour should differ. You can try to negate the logic
+  check.
+* Consider a corresponding Tumbleweed test: Because of
+  [Factory First](https://opensource.suse.com/suse-open-source-policy)
+  we should test against Tumbleweed by default (with exceptions for special
+  packages, of course). So please check if your code changes work with a current
+  Tumbleweed snapshot and then add to the according schedule for Tumbleweed
+  tests, e.g. main.pm
+* Avoid "dead code": Don't add disabled code as nobody but you will understand
+  why this is not active. Better leave it out and keep in your local git
+  repository, either in `git stash` or a temporary "WIP"-commit.
+* Details in commit messages: The commit message should have enough details,
```

Comment:
Hm, I'm a bit torn on this one to be honest. While `git blame` is very effective at showing you past commits, it fails pretty quickly once functions get moved around. Then you quickly loose all the past context and so I think that information relevant to the code itself **must** stay in the actual code.

Reply:
I personally don't have much faith for commit history and fully support idea of putting more info into  comments in code , function/variable/class names etc. . But on other hand I don't see any contradiction with what written here . Having some info in commit message should not stop you from doing something else to improve general knowledge about codebase

Reply:
I'm actually a big fan having having meaningful git commit message (some people put a lot of info into PR text box, which will be lost if we migrate from github to some other git hosting).

Reply:
@dcermak I agree that if the source can not explain itself then comments in the source should describe the choices made, to answer the question "why is the code like this". The git commit message should describe "why did we change it" which I don't see as contradiction.

Reply:
@okurz I don't disagree with you at all here. What I think could be clearer here though, is what should go into the commit message and what should be added as a comment (and sometimes it even makes sense to put some historical info into the comments, as very few people actually do a `git blame` before they start fixing "obvious" issues in a test that were added for a reason).

Reply:
Ok, How about an additional rule just after the above one:

```
* Add comments to the source code if the code is not self-explanatory:
  Comments in the source code should describe the choices made, to answer the
  question "why is the code like this". The git commit message should describe
  "why did we change it".
```

I added that now in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/11546/commits/af4f93d4757cb556a4f3a14a2ba01e78379031ae